cl/citations/match_citations.py

```diff
@@ -139,3 +146,110 @@ def match_citation(citation, citing_doc=None):
+        matched_opinion = None
+
+        # If the citation is a supra citation, try to resolve it to one of
+        # the citations that has already been matched
+        if isinstance(citation, SupraCitation):
+            for cm in citation_matches:
+                # The only clue we have to help us with resolution is the guess
+                # of what the supra citation's antecedent is, so we try to
+                # match that string to one of the known case names of the
+                # already matched opinions.
+                if (
+                    strip_punct(citation.antecedent_guess)
+                    in cm.cluster.case_name_full
+                ):
+                    # Just use the first one found, since the candidates should
+                    # all be identical. If nothing is found, then the supra
+                    # reference is effectively dropped.
+                    matched_opinion = cm
+                    break
+
+        # Likewise, if the citation is a short form citation, try to resolve it
+        # to one of the citations that has already been matched
+        elif isinstance(citation, ShortformCitation):
+            # We first try to match by using the reporter and volume number.
+            # However, because matches made using this heuristic may not be
+            # unique, we then refine by using the antecedent guess and only
+            # accept the match if there is a single unique candidate. This
+            # refinement may still fail (because the guess could be
+            # meaningless), in which case the citation is not resolvable and
+            # is dropped.
+            candidates = []
+            for cm in citation_matches:
+                for c in cm.cluster.citations.all():
+                    if (
+                        citation.reporter == c.reporter
+                        and citation.volume == c.volume
+                    ):
+                        candidates.append(cm)
+
+            candidates = list(set(candidates))  # Remove duplicate matches
+            if len(candidates) == 1:
+                matched_opinion = candidates[0]  # Accept the match!
+            else:
+                refined_candidates = []
+                for cm in candidates:
+                    if (
+                        strip_punct(citation.antecedent_guess)
+                        in cm.cluster.case_name_full
+                    ):
+                        refined_candidates.append(cm)
+
+                if len(refined_candidates) == 1:
+                    matched_opinion = refined_candidates[
+                        0
+                    ]  # Accept the match!
+
+        # If the citation is an id citation, just resolve it to the opinion
+        # that was matched immediately prior (so long as the previous match
+        # was successful).
+        elif isinstance(citation, IdCitation):
+            if was_matched:
+                matched_opinion = citation_matches[-1]
```

Comment:
These negative indexes are gonna bite us. Will this crash if the first citation we find is an `IdCitation`? I haven't done so carefully, but I suspect a careful pass over the PR looking at negative indexes would find some other instances too.

Reply:
It shouldn't crash here, because `was_matched` is initialized to `False`. So even if we find an `IdCitation` first, the code knows there's nothing in the `citation_matches` array to try to grab. However, I'll add a test case for this and review the negative indexing logic elsewhere.

Reply:
1. Good point. 

2. Thanks.

Reply:
Fixed in c839569. This was the only instance of us doing negative indexing in this file; I addressed other possible problems in d024203.